## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [6]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [14]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.columns .values. tolist() 

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [5]:
# What three item categories had the most entries?
df["category"].value_counts(ascending= False).head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
df.groupby(['category', 'subcategory'])["category"].count().sort_index().head(1)

category     subcategory      
consumables  bathroom supplies    6424
Name: category, dtype: int64

In [11]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts(ascending= False).head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [18]:
# Store the client ids of those top 5 clients in a list.
top_client_list=df["client_id"].value_counts(ascending= False).head(5).index.tolist()
top_client_list

[33615, 66037, 46820, 38378, 24741]

In [20]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_list = df["client_id"].value_counts().head(1).index.tolist()
cond1 = "client_id == " + str(top_client_list[0])
df.query(cond1)['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [5]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
def sub_total(row):  
    return  row['unit_price'] * row['qty']
 
df['row_subtotal'] = df.apply(lambda row: sub_total(row), axis=1)

print(df[['unit_price', 'qty', 'row_subtotal']].head(10))


   unit_price  qty  row_subtotal
0     1096.80  105     115164.00
1       24.95   21        523.95
2       13.52   39        527.28
3       36.42   29       1056.18
4      195.10   20       3902.00
5      172.84   78      13481.52
6      245.56   56      13751.36
7       20.99   22        461.78
8       83.86   49       4109.14
9       66.88   78       5216.64


In [6]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def ship_price(row):  
    item_weight = row['unit_weight'] * row['qty']
    if item_weight > 50:
        shipping_price = item_weight * 7
    else:
        shipping_price = item_weight *10
    return shipping_price 
df['shipping_price'] = df.apply(lambda row: ship_price(row), axis=1)

print(df[['unit_price', 'unit_weight', 'qty', 'shipping_price']].head(10))


   unit_price  unit_weight  qty  shipping_price
0     1096.80         7.50  105         5512.50
1       24.95         1.49   21          312.90
2       13.52         1.68   39          458.64
3       36.42         1.23   29          356.70
4      195.10        46.43   20         6500.20
5      172.84         2.79   78         1523.34
6      245.56         1.73   56          678.16
7       20.99         5.12   22          788.48
8       83.86         0.27   49          132.30
9       66.88         2.47   78         1348.62


In [7]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
def tot_price(row):  
    item_price = row['row_subtotal'] + row['shipping_price'] + ((row['row_subtotal'] + row['shipping_price']) * .0925) 
    return round(item_price  ,2) 

df['total_price'] = df.apply(lambda row: tot_price(row), axis=1)

print(df[['row_subtotal', 'shipping_price', 'total_price']].head(10))

   row_subtotal  shipping_price  total_price
0     115164.00         5512.50    131839.08
1        523.95          312.90       914.26
2        527.28          458.64      1077.12
3       1056.18          356.70      1543.57
4       3902.00         6500.20     11364.40
5      13481.52         1523.34     16392.81
6      13751.36          678.16     15764.25
7        461.78          788.48      1365.91
8       4109.14          132.30      4633.77
9       5216.64         1348.62      7172.55


In [75]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

def cal_cost(row):  
    cost = (row['unit_cost'] * row['qty'])+ row['shipping_price'] 
    return round(cost,2) 
df['cost'] = df.apply(lambda row: cal_cost(row), axis=1)
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ... unit_cost unit_weight  \
0   8953482  2023-04-28          17        2023  ...    762.71        7.50   
1   8069089  2023-05-19          20        2023  ...     15.09        1.49   
2   1902144  2023-01-29           4        2023  ...      7.86        1.68   
3   9031802  2023-04-25          17        2023  ...     24.85        1.23   
4   1322274  2023-05-28          21        2023  ...    108.17       46.43   
5   1327730  2023-05-02          18        2023  ...    154.44        2.79   
6   8080280  2023-04-19          16        2023  ...    231.95        1.73   
7   8946782  2023-02-14           7        2023  ...     16.97        5.12   
8   9394118  2023-05-13          19        2023  ...     59.00        0.27   
9   9665852  2023-04-14          15        2023  ...     54.15        2.47   

   qty  line_number  row_subtotal  shipping_price  total_price      cost  \
0  105            1     115164.00         5512.50    131839.08  85597.05   
1   21            0        523.95          312.90       914.26    629.79   
2   39            6        527.28          458.64      1077.12    765.18   
3   29            3       1056.18          356.70      1543.57   1077.35   
4   20            1       3902.00         6500.20     11364.40   8663.60   
5   78            0      13481.52         1523.34     16392.81  13569.66   
6   56            5      13751.36          678.16     15764.25  13667.36   
7   22            4        461.78          788.48      1365.91   1161.82   
8   49            3       4109.14          132.30      4633.77   3023.30   
9   78            1       5216.64         1348.62      7172.55   5572.32   

   line_profit    profit  
0     46242.03  46242.03  
1       284.47    284.47  
2       311.94    311.94  
3       466.22    466.22  
4      2700.80   2700.80  
5      2823.15   2823.15  
6      2096.89   2096.89  
7       204.09    204.09  
8      1610.47   1610.47  
9      1600.23   1600.23  

[10 rows x 24 columns]

In [76]:
# Create a column for the profit of each line using line cost and line price
def cal_profit(row):  
    profit = (row['total_price'] - row['cost'])
    return profit 

df['profit'] = df.apply(lambda row: cal_profit(row), axis=1)
df.head(10)


first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ... unit_cost unit_weight  \
0   8953482  2023-04-28          17        2023  ...    762.71        7.50   
1   8069089  2023-05-19          20        2023  ...     15.09        1.49   
2   1902144  2023-01-29           4        2023  ...      7.86        1.68   
3   9031802  2023-04-25          17        2023  ...     24.85        1.23   
4   1322274  2023-05-28          21        2023  ...    108.17       46.43   
5   1327730  2023-05-02          18        2023  ...    154.44        2.79   
6   8080280  2023-04-19          16        2023  ...    231.95        1.73   
7   8946782  2023-02-14           7        2023  ...     16.97        5.12   
8   9394118  2023-05-13          19        2023  ...     59.00        0.27   
9   9665852  2023-04-14          15        2023  ...     54.15        2.47   

   qty  line_number  row_subtotal  shipping_price  total_price      cost  \
0  105            1     115164.00         5512.50    131839.08  85597.05   
1   21            0        523.95          312.90       914.26    629.79   
2   39            6        527.28          458.64      1077.12    765.18   
3   29            3       1056.18          356.70      1543.57   1077.35   
4   20            1       3902.00         6500.20     11364.40   8663.60   
5   78            0      13481.52         1523.34     16392.81  13569.66   
6   56            5      13751.36          678.16     15764.25  13667.36   
7   22            4        461.78          788.48      1365.91   1161.82   
8   49            3       4109.14          132.30      4633.77   3023.30   
9   78            1       5216.64         1348.62      7172.55   5572.32   

   line_profit    profit  
0     46242.03  46242.03  
1       284.47    284.47  
2       311.94    311.94  
3       466.22    466.22  
4      2700.80   2700.80  
5      2823.15   2823.15  
6      2096.89   2096.89  
7       204.09    204.09  
8      1610.47   1610.47  
9      1600.23   1600.23  

[10 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
orderID_2742071 = round(df.loc[df['order_id'] == 2742071, 'total_price'].sum(),2)
print("Order ID 2742071 had a total price of $"+f"{orderID_2742071:,}")

orderID_2173913 = round(df.loc[df['order_id'] == 2173913, 'total_price'].sum(),2)
print("Order ID 2173913 had a total price of $"+f"{orderID_2173913:,}")

orderID_6128929 = round(df.loc[df['order_id'] == 6128929, 'total_price'].sum(),2)
print("Order ID 6128929 had a total price of $"+f"{orderID_6128929:,}")

Order ID 2742071 had a total price of $152,811.89
Order ID 2173913 had a total price of $162,388.71
Order ID 6128929 had a total price of $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

for top_client in  df["client_id"].value_counts().head(5).index.tolist():
    print(top_client,':',f"{round(df.loc[df['client_id'] == top_client, 'total_price'].sum(),2):}")

33615 : 8377308.52
66037 : 10259514.8
46820 : 9743794.33
38378 : 12906550.87
24741 : 82268892.04


In [109]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_df = df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('total_price', 'sum'),
    total_revenue=('cost', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()
top_clients_df = summary_df.sort_values(by='total_revenue', ascending=False).head(5)
top_clients_df


client_id  total_units_purchased  total_shipping_price  total_revenue  \
758      78965                4033302          6.559237e+08   4.085708e+08   
320      38473                 402770          1.835526e+08   1.402008e+08   
850      86721                1010426          1.812663e+08   1.382916e+08   
840      85931                 644423          2.374170e+08   1.281229e+08   
585      62227                 224149          1.002142e+08   8.459065e+07   

     total_profit  
758  2.473528e+08  
320  4.335173e+07  
850  4.297470e+07  
840  1.092941e+08  
585  1.562353e+07

In [110]:
# Format the data and rename the columns to names suitable for presentation.
top_clients_df = top_clients_df.rename(columns={'client_id':'Client ID', \
                                                'total_units_purchased':'Units',\
                                                'total_shipping_price':'Shipping (millions)',\
                                                    'total_revenue':'Total Revenue (millions)',\
                                                        'total_profit':'Total Profit (millions)'})
# Define the money columns. 

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return value / 1000000


# Apply the currency_format_millions function to only the money columns. 
top_clients_df['Shipping (millions)'] = top_clients_df['Shipping (millions)'].apply(currency_format_millions)
top_clients_df['Total Revenue (millions)'] = top_clients_df['Total Revenue (millions)'].apply(currency_format_millions)
top_clients_df['Total Profit (millions)'] = top_clients_df['Total Profit (millions)'].apply(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
#top_clients_df = top_clients_df.rename(columns={
#    'Shipping (millions)': 'Shipping (in millions)',
 #   'Total Revenue (millions)': 'Total Revenue (in millions)',
 #   'Total Cost (millions)': 'Total Cost (in millions)'
#})

top_clients_df.head(10)



Client ID    Units  Shipping (millions)  Total Revenue (millions)  \
758      78965  4033302           655.923670                408.570828   
320      38473   402770           183.552572                140.200845   
850      86721  1010426           181.266297                138.291602   
840      85931   644423           237.417001                128.122873   
585      62227   224149           100.214179                 84.590650   

     Total Profit (millions)  
758               247.352842  
320                43.351727  
850                42.974695  
840               109.294129  
585                15.623530

In [111]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
profit_sorted_df = top_clients_df.sort_values(by='Total Profit (millions)', ascending=False) 
profit_sorted_df.head(10)


Client ID    Units  Shipping (millions)  Total Revenue (millions)  \
758      78965  4033302           655.923670                408.570828   
840      85931   644423           237.417001                128.122873   
320      38473   402770           183.552572                140.200845   
850      86721  1010426           181.266297                138.291602   
585      62227   224149           100.214179                 84.590650   

     Total Profit (millions)  
758               247.352842  
840               109.294129  
320                43.351727  
850                42.974695  
585                15.623530